### DISCUSSION TITLE
deletate*<...> is the same as deletate* managed<...> and why it is unsafe ?

### DISCUSSION THREAD

**Discussion by redradist** [dotnet/csharplang#4862] [General] [answered]
I have learned latest feature from C# 9 `function pointer` and I have one strange question ...
Why `deletate*<...>` is unsafe ?

Only because it is pointer ? But it is pointer to managed function it is similar to delegate with one callback ...
I would consider `unsafe` only the following pointer `deletate* unmanaged<...>`

Is there an answer ?
Or because such construct is called pointer I need to add to project unsafe keyword ...
`deletate*<...>` is usefull in cases when I want to be sure that user pass only single function in delegate in contrast with general delegate (which are actually MultiDelegate)

**Comment by 333fred** [comment, marked as answer]
It's unsafe because the GC doesn't track the pointer. If you take a pointer to a method in a different assembly and that assembly is later unloaded, nothing invalidates that pointer. It's now just a dangling pointer to undefined and what happens when you call it could be anything from the code working to formatting your hard drive.

**Reply by redradist** [reply]
But why there are no single delegate object that can take only single managed method without worring of unloaded dll ?
For example with such syntax:
```csharp
delegate<int , int, int> PerformCalculation;
```

**Reply by 333fred** [reply]
It was considered, but ultimately what we can implement on the CLR in this space today is just regular delegates with a different name. Several members of the LDM are interested in more lightweight function references (what we've been terming "value delegates"), but that work is predicated on several other features such as ref structs in generics and ref structs being able to implement interfaces.

**Reply by redradist** [reply]
I will argue that we cannot do it today ...
I mean, I cannot enforce user to pass only on callback function and check it in compile time ... what I can do is that to check how many functions user passed to me and if there are more than 1 function I can throw exception ...
But runtime failure just mean that some feature is missed in language and I cannot do it in compile time

Anyway this feature with single method as delegate would be very helpful

**Reply by 333fred** [reply]
You could potentially write an analyzer, though it would certainly not be perfect. I also don't particularly understand the motivation here: why do you care whether the user passes in a multicast delegate with multiple targets?

**Reply by huoyaoyuan** [reply]
> Several members of the LDM are interested in more lightweight function references (what we've been terming "value delegates")

I'm so glad to hear this. I was always requiring CLR support for this before considering in language.
